federation/executor.go

```diff
@@ -59,72 +95,52 @@ func fetchSchema(ctx context.Context, e ExecutorClient, metadata interface{}) (*
+type CustomExecutorArgs struct {
+	schemaSyncer             SchemaSyncer
+	optionalArgs             interface{}
+	schemaSyncIntervalSecond int64
 }
 
-func NewExecutorWithOptionalArgs(ctx context.Context, executors map[string]ExecutorClient, optionalArgs interface{}) (*Executor, error) {
-	// Fetches the schemas from the executors clients
-	schemas := make(map[string]*introspectionQueryResult)
-	for server, client := range executors {
-		resp, err := fetchSchema(ctx, client, optionalArgs)
-		schema := resp.Result
-		if err != nil {
-			return nil, oops.Wrapf(err, "fetching schema %s", server)
-		}
-
-		var iq introspectionQueryResult
-		if err := json.Unmarshal(schema, &iq); err != nil {
-			return nil, oops.Wrapf(err, "unmarshaling schema %s", server)
-		}
-
-		schemas[server] = &iq
+func NewExecutor(ctx context.Context, executors map[string]ExecutorClient, c *CustomExecutorArgs) (*Executor, error) {
+	if c.schemaSyncer == nil {
+		c.schemaSyncer = newIntrospectionSchemaSyncer(ctx, executors, c.optionalArgs)
 	}
-
-	types, err := convertSchema(schemas)
-	if err != nil {
-		return nil, oops.Wrapf(err, "converting schemas error")
-	}
-
-	introspectionSchema := introspection.BareIntrospectionSchema(types.Schema)
-	introspectionServer := &Server{schema: introspectionSchema}
-
-	executors["introspection"] = &DirectExecutorClient{Client: introspectionServer}
-	schema, err := introspection.RunIntrospectionQuery(introspection.BareIntrospectionSchema(introspectionServer.schema))
-
-	var iq introspectionQueryResult
-	if err := json.Unmarshal(schema, &iq); err != nil {
-		return nil, oops.Wrapf(err, "unmarshaling introspection schema")
+	if c.schemaSyncIntervalSecond == 0 {
+		c.schemaSyncIntervalSecond = minSchemaSyncIntervalSeconds
 	}
-
-	schemas["introspection"] = &iq
-	types, err = convertSchema(schemas)
+	planner, err := c.schemaSyncer.FetchPlanner(ctx, c.optionalArgs)
 	if err != nil {
-		return nil, oops.Wrapf(err, "converting schemas error")
+		return nil, oops.Wrapf(err, "failed to load schema")
 	}
-
-	flattener, err := newFlattener(types.Schema)
-	if err != nil {
-		return nil, oops.Wrapf(err, "flattening schemas error")
+	executor := &Executor{
+		Executors: executors,
+		syncer: &Syncer{
+			ticker:       time.NewTicker(time.Duration(c.schemaSyncIntervalSecond) * time.Second),
+			schemaSyncer: c.schemaSyncer,
+			plannerMu:    &sync.RWMutex{},
+			planner:      planner,
+		},
 	}
+	go executor.poll(ctx, c.optionalArgs)
+	return executor, nil
+}
 
-	// The planner is aware of the merged schema and what executors
-	// know about what fields
-	planner := &Planner{
-		schema:    types,
-		flattener: flattener,
+func (e *Executor) poll(ctx context.Context, optionalArgs interface{}) {
+	for {
+		select {
+		case <-e.syncer.ticker.C:
+			newPlanner, err := e.syncer.schemaSyncer.FetchPlanner(ctx, optionalArgs)
+			if err == nil && newPlanner != nil {
+				e.setPlanner(newPlanner)
+			}
+		case <-ctx.Done():
```

Comment:
Should we return ctx.Err here?

Reply:
This function doesn't return anything, so I'm not really sure where the error will get reported? Its also run as a separate goroutine? Could you explain more about what we can do to catch the error?

Reply:
Then maybe just return for now?